src/core/device.c

```diff
@@ -158,59 +158,38 @@ static void device_set_state(Device *d, DeviceState state) {
-         *   other units, or (b) generated when uevents are received.
-         *
-         * - On switch-root, the udev database may be cleared, except for devices with sticky bit, i.e.
-         *   OPTIONS="db_persist". Hence, almost no devices are enumerated in the step 2. However, in general,
-         *   we have several serialized devices. So, DEVICE_FOUND_UDEV bit in the deserialized_found must be
-         *   ignored, as udev rules in initramfs and the main system are often different. If the deserialized
-         *   state is DEVICE_PLUGGED, we need to downgrade it to DEVICE_TENTATIVE (or DEVICE_DEAD if nobody
-         *   sees the device). Unlike the other starting mode, Manager.honor_device_enumeration == false
-         *   (maybe, it is better to rename the flag) when device_coldplug() and device_catchup() are called.
-         *   Hence, let's conditionalize the operations by using the flag. After switch-root, systemd-udevd
-         *   will (re-)process all devices, and the Device.found and Device.state will be adjusted.
-         *
-         * - On reload or reexecute, we can trust enumerated_found, deserialized_found, and deserialized_state.
-         *   Of course, deserialized parameters may be outdated, but the unit state can be adjusted later by
-         *   device_catchup() or uevents. */
-
-        if (!m->honor_device_enumeration && !MANAGER_IS_USER(m)) {
-                found &= ~DEVICE_FOUND_UDEV; /* ignore DEVICE_FOUND_UDEV bit */
-                if (state == DEVICE_PLUGGED)
-                        state = DEVICE_TENTATIVE; /* downgrade state */
-                if (found == DEVICE_NOT_FOUND)
-                        state = DEVICE_DEAD; /* If nobody sees the device, downgrade more */
-        }
+        found = d->deserialized_found;
+        state = d->deserialized_state;
 
-        if (d->found == found && d->state == state)
-                return 0;
+        if (!FLAGS_SET(d->enumerated_found, DEVICE_FOUND_UDEV) &&
+            MANAGER_IS_SWITCHING_ROOT(u->manager)) {
+
+                /* The device is no more present in udev DB. When this happens (due to PID1
+                 * switching root), udev is inactive and its DB has been cleared, hence we can't
+                 * trust the enumeration step anymore until udev DB is repopulated again. Until
+                 * this mechanism is used during initrd transitions, we have no other choice than
+                 * forgetting all devices and relying on a full retriggering of events (based on
+                 * the host set of udev rules this time). Hence we avoid generating wrong device
+                 * state transitions by not assuming wrongly that devices missing in udev DB were
+                 * unplugged. The downside is that PID1 won't generate plugged->dead transitions
+                 * for devices that were unplugged during the initrd transition. */
+
+                found = DEVICE_NOT_FOUND;
+                state = DEVICE_DEAD;
```

Comment:
I'm sorry but I don't get this. Can you please elaborate ?

Reply:
With this change, if a device does NOT have persistent database, then the device is going to dead state here, and later possibly goes to plugged state on uevent. Hence, if the device is already mounted on somewhere, it will be once unmounted. Thus, the original issue is re-introduced.

If a device HAS persistent databse, then the up-to-date flags (and state) will be applied to device_catchup(). Hence, it is not necessary to downgrade the flag and state here. So, the condition added by the second commit is correct.

So, a correct way to fix #23429 is simply add the condition.

Reply:
Consider the case that a device does NOT have persistent database, and it is mounted on somewhere.
With this PR, the device will enter the dead state, as `enumerated_found` does not have DEVICE_FOUND_UDEV flag when `device_coldplug()` is called. Hence, the corresponding mount unit will be stopped. After systemd-udevd processes the device, then the device unit again goes to the plugged state, and the mount unit is restarted.

This is mostly equivalent to the original issue, right?

Reply:
No.

If the device is mounted, then `d->enumerated_found=DEVICE_FOUND_MOUNT` and the device enters in "tentative state".

Then later udevd is started and the events are replayed. From this point, the device enters in "plugged" state through `tentative->plugged` transition, which doesn't stop the mount unit.

I think you should give it a try.